source/module_hamilt_lcao/module_tddft/upsi.cpp

```diff
@@ -40,43 +41,237 @@ void upsi(const Parallel_Orbitals* pv,
+
+template <typename Device>
+void upsi_tensor_lapack(const Parallel_Orbitals* pv,
+                        const int nband,
+                        const int nlocal,
+                        const ct::Tensor& U_operator,
+                        const ct::Tensor& psi_k_laststep,
+                        ct::Tensor& psi_k,
+                        std::ofstream& ofs_running,
+                        const int print_matrix)
+{
+    // ct_device_type = ct::DeviceType::CpuDevice or ct::DeviceType::GpuDevice
+    ct::DeviceType ct_device_type = ct::DeviceTypeToEnum<Device>::value;
+    // ct_Device = ct::DEVICE_CPU or ct::DEVICE_GPU
+    using ct_Device = typename ct::PsiToContainer<Device>::type;
+
+    // Perform the matrix multiplication: psi_k = U_operator * psi_k_laststep
+    std::complex<double> alpha = {1.0, 0.0};
+    std::complex<double> beta = {0.0, 0.0};
+
+    ct::kernels::blas_gemm<std::complex<double>, ct_Device>()('N',
+                                                              'N',
+                                                              nlocal,
+                                                              nband,
+                                                              nlocal,
+                                                              &alpha,
+                                                              U_operator.data<std::complex<double>>(),
+                                                              nlocal,
+                                                              psi_k_laststep.data<std::complex<double>>(),
+                                                              nlocal,
+                                                              &beta,
+                                                              psi_k.data<std::complex<double>>(),
+                                                              nlocal);
+
+    if (print_matrix)
+    {
+        ct::Tensor psi_k_cpu = psi_k.to_device<ct::DEVICE_CPU>();
+        ct::Tensor psi_k_laststep_cpu = psi_k_laststep.to_device<ct::DEVICE_CPU>();
+
+        ofs_running << std::endl;
+        ofs_running << " psi_k:" << std::endl;
+        for (int i = 0; i < nband; i++)
+        {
+            for (int j = 0; j < nlocal; j++)
+            {
+                double aa = 0.0;
+                double bb = 0.0;
+                aa = psi_k_cpu.data<std::complex<double>>()[i * nlocal + j].real();
+                bb = psi_k_cpu.data<std::complex<double>>()[i * nlocal + j].imag();
+                if (std::abs(aa) < 1e-8)
+                {
+                    aa = 0.0;
+                }
+                if (std::abs(bb) < 1e-8)
+                {
+                    bb = 0.0;
+                }
+                ofs_running << aa << "+" << bb << "i ";
+            }
+            ofs_running << std::endl;
+        }
+        ofs_running << std::endl;
+        ofs_running << " psi_k_laststep:" << std::endl;
+        for (int i = 0; i < nband; i++)
+        {
```

Comment:
const int in = i * nlocal; add this code in the first loop (i loop), and then directly use in+j in the second loop